vignettes/d3-to-shiny-communication.Rmd

```diff
@@ -0,0 +1,98 @@
+
+## Example
+
+In this example, the bar plot will pass to Shiny the data value of the bar plot that was clicked by the user.  Shiny will then display a notification containing the value of the bar that was clicked on.
+
+Here is the D3 script:
+
+```js
+// !preview r2d3 data=c(0.3, 0.6, 0.8, 0.95, 0.40, 0.20)
+
+var barHeight = Math.ceil(height / data.length);
+
+svg.selectAll('rect')
+  .data(data)
+  .enter().append('rect')
+    .attr('width', function(d) { return d * width; })
+    .attr('height', barHeight)
+    .attr('y', function(d, i) { return i * barHeight; })
+    .attr('fill', 'steelblue')
+    .attr("id", function(d) { return d; })
+    .on("click", function(){
+            Shiny.onInputChange(
+              "bar_clicked", 
+              d3.select(this).attr("id"));
+        });
+```
+
+### D3 Code breakdown
+
+First, we need to decide which attribute (`.attr`) is going to be passed to Shiny when the user clicks.  Any existing attribute can be passed, such as the value of `y` or `width`, but because no current attribute contain the raw value of the data, a new made up attribute is added called `id`.  The `id` attribute is set to rerun the `d` variable.
+
+<img src="images/attr.PNG" class="illustration" width=600/>
+
+Next, the `on()` function is used to capture an event inside the plot.  The first argument is "click", indicating that we want it to be activated when someone clicks on the plot.  The second argument is a function, this function will contain a call to the `Shiny.onInputChange()` JS function.  The first argument of the Shiny JS function is the name that the Shiny app will be looking for. Using a name describing what kind of action we are tracking will be preferred, makes it easier for Shiny developers that include this D3 plot to remember the name. In this case, we will use **bar_clicked**.  The second argument is the value that is returned to Shiny, and for that, we will use the `id` attribute we prepared in the previous code entry.
+
+<img src="images/onclick.PNG" class="illustration" width=600/>
+
+### Shiny
+
+Place an `observeEvent()` function that looks for a Shiny input named **bar_clicked**, which is the name we gave to the Shiny JS call inside the D3 script:
+
+```r
+observeEvent(input$bar_clicked,{  })
+```  
+
+This will only trigger an action when the bar is clicked, to know the value that was clicked on, use the exact same input inside the `observeEvent()` command.  For this example, we will just pass the value to a `showNotification()` function.
+
+```r
+    observeEvent(input$bar_clicked,{
+        showNotification(input$bar_clicked)
```

Comment:
I think we should have a demo that uses the input in a calculation (as that will be the most common thing for users to do). That will also force us to introduce the use of the `req()` function (since the input will be `NULL` until the user clicks)